dataplane/standalone/saiserver/attrmgr/attrmgr.go

```diff
@@ -0,0 +1,155 @@
+package attrmgr
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protowire"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+
+	saipb "github.com/openconfig/lemming/dataplane/standalone/proto"
+)
+
+// AttrMgr stores and retrieve the SAI attributes.
+type AttrMgr struct {
+	attrsMu sync.RWMutex
+	attrs   map[string]map[int32]protoreflect.Value
+	nextOid atomic.Uint64
+}
+
+// New returns a new AttrMgr.
+func New() *AttrMgr {
+	mgr := &AttrMgr{
+		attrs: make(map[string]map[int32]protoreflect.Value),
+	}
+	return mgr
+}
+
+func (mgr *AttrMgr) set(id string, attr int32, val protoreflect.Value) {
+	mgr.attrsMu.Lock()
+	defer mgr.attrsMu.Unlock()
+	if _, ok := mgr.attrs[id]; !ok {
+		mgr.attrs[id] = make(map[int32]protoreflect.Value)
+	}
+	mgr.attrs[id][attr] = val
+}
+
+var (
+	editRPC = regexp.MustCompile("Create|Set")
+	readRPC = regexp.MustCompile("Get")
+)
+
+const protoNS = "lemming.dataplane.sai"
+
+// Interceptor returns a gRPC interceptor that automatically store values set in requests and fills in responses with the stored values.
+func (mgr *AttrMgr) Interceptor(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+	if !strings.Contains(info.FullMethod, protoNS) {
+		return handler(ctx, req)
+	}
+	resp, err := handler(ctx, req)
+	// Ignore unimplemented errors, so that we don't have to implement APIs we don't support.
+	if st, _ := status.FromError(err); err != nil && (st.Code() != codes.Unimplemented) {
+		return resp, err
+	}
+	reqMsg := req.(proto.Message)
+	// If the resp is nil, then create a new response type.
+	if resp == nil {
+		respName := strings.ReplaceAll(string(reqMsg.ProtoReflect().Descriptor().FullName()), "Request", "Response")
+		pType, err := protoregistry.GlobalTypes.FindMessageByName(protoreflect.FullName(respName))
+		if err != nil {
+			return nil, err
+		}
+		resp = pType.New().Interface()
+	}
+	respMsg := resp.(proto.Message)
+	if res := editRPC.FindString(info.FullMethod); res != "" {
+		id, err := mgr.getID(reqMsg, respMsg)
+		if err != nil {
+			return nil, err
+		}
+		mgr.storeAttributes(id, reqMsg)
+	} else if res := readRPC.FindString(info.FullMethod); res != "" {
+		id, err := mgr.getID(reqMsg, respMsg)
+		if err != nil {
+			return nil, err
+		}
+		mgr.populateAttributes(id, reqMsg, respMsg)
+	}
+	return resp, nil
+}
+
+// populateAttributes fills the resp with the requests attributes.
+func (mgr *AttrMgr) populateAttributes(id string, req, resp proto.Message) {
+	mgr.attrsMu.RLock()
+	defer mgr.attrsMu.RUnlock()
+
+	attrs := resp.ProtoReflect().Mutable(resp.ProtoReflect().Descriptor().Fields().ByTextName("attr")).Message()
```

Comment:
I think a comment here about what types of protobufs you are expecting here would be helpful.  Also, what happens if the protobuf is random and not what you expect?

Reply:
add comment and check for this